lib/event_store/sql/statements/insert_events.sql.eex

```diff
@@ -1,15 +1,76 @@
-INSERT INTO <%= schema %>.events
-  (
-    event_id,
-    event_type,
-    causation_id,
-    correlation_id,
-    data,
-    metadata,
-    created_at
+WITH
+  new_events_indexes (event_id, index, stream_version) AS (
+    VALUES
+    <%= for i <- 0..(number_of_events - 1) do %>
+      <%= unless i == 0 do %>,<% end %>
+      ($<%= i*9+3 %>::uuid, $<%= i*9+10 %>::int, $<%= i*9+11 %>::bigint)
+    <% end %>
+  ),
+  events AS (
+    INSERT INTO <%= schema %>.events
+      (
+        event_id,
+        event_type,
+        causation_id,
+        correlation_id,
+        data,
+        metadata,
+        created_at
+      )
+    VALUES
+    <%= for i <- 0..(number_of_events - 1) do %>
+      <%= unless i == 0 do %>,<% end %>
+      ($<%= i*9+3 %>, $<%= i*9+4 %>, $<%= i*9+5 %>, $<%= i*9+6 %>, $<%= i*9+7 %>, $<%= i*9+8 %>, $<%= i*9+9 %>)
```

Comment:
For readers, can you explain these magic numbers? I would have guessed it was the positional argument, but there's 7?

Reply:
The magic numbers relate to the positions of the fields from the events to be inserted. In addition `$1` is the stream id or stream uuid and `$2` is the event count. The 9 is used as the offset for each event's fields (7 fields per event + `$1` and `$2`).

Reply:
The meanings of the variables is found on the lines above, `INSERT INTO (a, b, c) VALUES ($1, $2, $3), ($4, $5, $6)` is essentially what's happening here, where `a` corresponds to `$1`, and `$4`, and so on. This is just the same thing but a much longer list.

Reply:
Thanks for the explanation